From 64115832bea2994422ac3302d07dcf732ffd9f1b Mon Sep 17 00:00:00 2001
From: Dan Liang <dan.liang@intel.com>
Date: Fri, 17 May 2013 11:22:13 +0800
Subject: [PORT FROM MAIN] To fix the out-of-memory problem when playing video in VideoEditor

BZ: 74638

Root cause:
1. In JB MR1, the ConsumerBase class is introduced, and it's used as
the base class for SurfaceTexture.

2. There is a virtual function called freeLockedBuffer in both
ConsumerBase and SurfaceTexture.

3. When a SurfaceTexture instance is deconstructed, the ConsumerBase
deconstrutor is called. In ~ConsumerBase, freeLockedBuffer() is called.
The author's intent is to call the freeLockedBuffer() in the derived
SurfaceTexture class. However, according to the C++ language,
only the freeLockedBuffer() in ConsumerBase is called.

4. This leads to wrong GraphicBuffer reference counts. When a
SurfaceTexture instance is destroyed in video editor, the GraphicBuffer
decontructor is not called. The allocated GraphicBuffer instances are not
freed correctly, which causes the sharedbuffer in the Gralloc HAL
out-of-memory...

The solution is that use the SurfaceTexture decontructor to all
the SurfaceTexture::freeLockedBuffer().

Change-Id: Ibc7be92a0824248e0af1283f46a80f3116f89e53
Orig-Change-Id: Ia2477100254d28ca2f63aa09651e8eee8b601307
Signed-off-by: wfeng6 <wei.feng@intel.com>
Signed-off-by: Dan Liang <dan.liang@intel.com>
---
 include/gui/GLConsumer.h |    2 ++
 libs/gui/GLConsumer.cpp  |    5 +++++
 2 files changed, 7 insertions(+), 0 deletions(-)

diff --git a/include/gui/GLConsumer.h b/include/gui/GLConsumer.h
index f0a75dc..cbe6f05 100644
--- a/include/gui/GLConsumer.h
+++ b/include/gui/GLConsumer.h
@@ -89,6 +89,8 @@ public:
             GLenum texTarget = GL_TEXTURE_EXTERNAL_OES, bool useFenceSync = true,
             const sp<BufferQueue> &bufferQueue = 0);
 
+    virtual ~GLConsumer();
+
     // updateTexImage acquires the most recently queued buffer, and sets the
     // image contents of the target texture to it.
     //
diff --git a/libs/gui/GLConsumer.cpp b/libs/gui/GLConsumer.cpp
index 637a403..0812d3e 100644
--- a/libs/gui/GLConsumer.cpp
+++ b/libs/gui/GLConsumer.cpp
@@ -104,6 +104,11 @@ GLConsumer::GLConsumer(GLuint tex, bool allowSynchronousMode,
     mBufferQueue->setConsumerUsageBits(DEFAULT_USAGE_FLAGS);
 }
 
+GLConsumer::~GLConsumer() {
+    ST_LOGV("~GLConsumer");
+    abandon();
+}
+
 status_t GLConsumer::setDefaultMaxBufferCount(int bufferCount) {
     Mutex::Autolock lock(mMutex);
     return mBufferQueue->setDefaultMaxBufferCount(bufferCount);
-- 
1.7.4.1

